File: CS/ODataService/XpoOdataService.svc.cs
using System;
using System.Collections.Generic;
using System.Data.Services;
using Northwind;
using DevExpress.Xpo;
using DevExpress.Xpo.DB;
using System.ServiceModel;
using System.IO;
using System.Web;
using System.Data.Services.Common;
using System.Linq.Expressions;
using System.ServiceModel.Web;
using System.Linq;
using System.Web.Security;

namespace XpoODataExample {
    public class XpoODataContext : XpoContext {
        public bool CategoriesFiltering = false;
        public XpoODataContext(string s1, string s2, IDataLayer dataLayer)
            : base(s1, s2, dataLayer) {
        }
        [Action]
        public bool CanRemove(Products product) {
            return !(new XPQuery<Suppliers>(product.Session)).Any(s => s.Products.Contains(product));
        }
    }

    [ServiceBehavior(InstanceContextMode = InstanceContextMode.PerCall)]
    public class XpoODataExampleService : XpoDataServiceV3 {
        static readonly XpoODataContext context = new XpoODataContext("XpoExampleContext", "Northwind", CreateDataLayer());

        public XpoODataExampleService() : base(context, context) { }

        static IDataLayer CreateDataLayer() {
            DevExpress.Xpo.Metadata.XPDictionary dict = new DevExpress.Xpo.Metadata.ReflectionDictionary();
            // Initialize the XPO dictionary. 
            dict.GetDataStoreSchema(typeof(Orders).Assembly);
            InMemoryDataStore store = new InMemoryDataStore(AutoCreateOption.SchemaOnly);
            string DBFileName = DevExpress.Utils.FilesHelper.FindingFileName(HttpRuntime.AppDomainAppPath, "App_Data\\nwind.xml");
            if(DBFileName != "")
                store.ReadXml(DBFileName);
            IDataLayer dataLayer = new ThreadSafeDataLayer(dict, store);
            XpoDefault.DataLayer = dataLayer;
            XpoDefault.Session = null;
            return dataLayer;
        }
        protected override void OnStartProcessingRequest(ProcessRequestArgs args) {
            base.OnStartProcessingRequest(args);
        }
        public static void InitializeService(DataServiceConfiguration config) {
            config.SetEntitySetAccessRule("*", EntitySetRights.All);
            config.DataServiceBehavior.AcceptAnyAllRequests = true;
            config.UseVerboseErrors = true;
            config.SetServiceOperationAccessRule("*", ServiceOperationRights.All);
            config.SetServiceActionAccessRule("*", ServiceActionRights.Invoke);
            config.DataServiceBehavior.MaxProtocolVersion = DataServiceProtocolVersion.V3;
            config.DataServiceBehavior.AcceptProjectionRequests = true;
            config.DataServiceBehavior.AcceptCountRequests = true;
            config.AnnotationsBuilder = CreateAnnotationsBuilder(() => context);
            config.DataServiceBehavior.AcceptReplaceFunctionInQuery = true;
            config.DataServiceBehavior.AcceptSpatialLiteralsInQuery = true;
            config.DisableValidationOnMetadataWrite = true;
        }

        [QueryInterceptor("Categories")]
        public Expression<Func<Categories, bool>> OnQueryActors() {
            if(context.CategoriesFiltering)
                return o => o.CategoryName.Length > 12;
            else
                return o => true;
        }

        [ChangeInterceptor("Products")]
        public void OnChangeProducts(Products product, UpdateOperations operations) {
            if (Equals(operations, UpdateOperations.Change)) {
                Products productInBase = GetEntityUpdated<Products>(product);
                if (productInBase.ProductName != product.ProductName) {
                    throw new DataServiceException(400, string.Format("A {0} cannot be modified.", product.ToString()));
                }
            } else if (Equals(operations, UpdateOperations.Delete)) {
                throw new DataServiceException(400, "Products cannot be deleted.");
            }
        }

        public override object Authenticate(ProcessRequestArgs args) {
            int oid;
            try {
                string authCookie = args.OperationContext.RequestHeaders["userCookie"];
                FormsAuthenticationTicket ticket = FormsAuthentication.Decrypt(authCookie);
                oid = int.Parse(ticket.UserData);
            } catch {
                return null;
            }
            using(UnitOfWork uow = new UnitOfWork(Context.ObjectLayer)) {
                Employees cont = uow.GetObjectByKey<Employees>(oid);
                return cont != null ? "ok" : ":(";
            }
        }

        public override LambdaExpression GetQueryInterceptor(Type entityType, object token) {
            if(token == null) throw new DataServiceException(401, "Unauthorized access.");
            if((string)token == "ok")
                return (Expression<Func<object, bool>>)(o => true);
            else {
                return (Expression<Func<object, bool>>)(o => false);
            }
        }

        [WebGet]
        public string GetAuthorizationCookie(string name, string password) {
            using(UnitOfWork uow = new UnitOfWork(Context.ObjectLayer)) {
                List<Employees> contacts = new XPQuery<Employees>(uow).Where(i => i.LastName == name).ToList();
                Employees contact = contacts.SingleOrDefault();

                if(contact == null) return string.Empty;
                FormsAuthenticationTicket ticket = new FormsAuthenticationTicket(
                    1,
                    contact.LastName,
                    DateTime.Now,
                    DateTime.Now.AddMinutes(30),
                    true,
                    contact.EmployeeID.ToString());

                string encryptedTicket = FormsAuthentication.Encrypt(ticket);
                return encryptedTicket;
            }
        }

        [WebGet]
        public void SetCategoriesFiltering(bool state) {
            context.CategoriesFiltering = state;
        }
    }
}

File: VB/ODataService/XpoOdataService.svc.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Data.Services
Imports Northwind
Imports DevExpress.Xpo
Imports DevExpress.Xpo.DB
Imports System.ServiceModel
Imports System.IO
Imports System.Web
Imports System.Data.Services.Common
Imports System.Linq.Expressions
Imports System.ServiceModel.Web
Imports System.Linq
Imports System.Web.Security

Namespace XpoODataExample
	Public Class XpoODataContext
		Inherits XpoContext
		Public CategoriesFiltering As Boolean = False
		Public Sub New(ByVal s1 As String, ByVal s2 As String, ByVal dataLayer As IDataLayer)
			MyBase.New(s1, s2, dataLayer)
		End Sub
		<Action> _
		Public Function CanRemove(ByVal product As Products) As Boolean
			Return Not(New XPQuery(Of Suppliers)(product.Session)).Any(Function(s) s.Products.Contains(product))
		End Function
	End Class

	<ServiceBehavior(InstanceContextMode := InstanceContextMode.PerCall)> _
	Public Class XpoODataExampleService
		Inherits XpoDataServiceV3
		Private Shared ReadOnly context As New XpoODataContext("XpoExampleContext", "Northwind", CreateDataLayer())

		Public Sub New()
			MyBase.New(context, context)
		End Sub

		Private Shared Function CreateDataLayer() As IDataLayer
			Dim dict As DevExpress.Xpo.Metadata.XPDictionary = New DevExpress.Xpo.Metadata.ReflectionDictionary()
			' Initialize the XPO dictionary. 
			dict.GetDataStoreSchema(GetType(Orders).Assembly)
			Dim store As New InMemoryDataStore(AutoCreateOption.SchemaOnly)
			Dim DBFileName As String = DevExpress.Utils.FilesHelper.FindingFileName(HttpRuntime.AppDomainAppPath, "App_Data\nwind.xml")
			If DBFileName <> "" Then
				store.ReadXml(DBFileName)
			End If
			Dim dataLayer As IDataLayer = New ThreadSafeDataLayer(dict, store)
			XpoDefault.DataLayer = dataLayer
			XpoDefault.Session = Nothing
			Return dataLayer
		End Function
		Protected Overrides Sub OnStartProcessingRequest(ByVal args As ProcessRequestArgs)
			MyBase.OnStartProcessingRequest(args)
		End Sub
		Public Shared Sub InitializeService(ByVal config As DataServiceConfiguration)
			config.SetEntitySetAccessRule("*", EntitySetRights.All)
			config.DataServiceBehavior.AcceptAnyAllRequests = True
			config.UseVerboseErrors = True
			config.SetServiceOperationAccessRule("*", ServiceOperationRights.All)
			config.SetServiceActionAccessRule("*", ServiceActionRights.Invoke)
			config.DataServiceBehavior.MaxProtocolVersion = DataServiceProtocolVersion.V3
			config.DataServiceBehavior.AcceptProjectionRequests = True
			config.DataServiceBehavior.AcceptCountRequests = True
			config.AnnotationsBuilder = CreateAnnotationsBuilder(Function() context)
			config.DataServiceBehavior.AcceptReplaceFunctionInQuery = True
			config.DataServiceBehavior.AcceptSpatialLiteralsInQuery = True
			config.DisableValidationOnMetadataWrite = True
		End Sub

		<QueryInterceptor("Categories")> _
		Public Function OnQueryActors() As Expression(Of Func(Of Categories, Boolean))
			If context.CategoriesFiltering Then
				Return Function(o) o.CategoryName.Length > 12
			Else
				Return Function(o) True
			End If
		End Function

		<ChangeInterceptor("Products")> _
		Public Sub OnChangeProducts(ByVal product As Products, ByVal operations As UpdateOperations)
			If Equals(operations, UpdateOperations.Change) Then
				Dim productInBase As Products = GetEntityUpdated(Of Products)(product)
				If productInBase.ProductName <> product.ProductName Then
					Throw New DataServiceException(400, String.Format("A {0} cannot be modified.", product.ToString()))
				End If
			ElseIf Equals(operations, UpdateOperations.Delete) Then
				Throw New DataServiceException(400, "Products cannot be deleted.")
			End If
		End Sub

		Public Overrides Function Authenticate(ByVal args As ProcessRequestArgs) As Object
			Dim oid As Integer
			Try
				Dim authCookie As String = args.OperationContext.RequestHeaders("userCookie")
				Dim ticket As FormsAuthenticationTicket = FormsAuthentication.Decrypt(authCookie)
				oid = Integer.Parse(ticket.UserData)
			Catch
				Return Nothing
			End Try
			Using uow As New UnitOfWork(Context.ObjectLayer)
				Dim cont As Employees = uow.GetObjectByKey(Of Employees)(oid)
				Return If(cont IsNot Nothing, "ok", ":(")
			End Using
		End Function

		Public Overrides Function GetQueryInterceptor(ByVal entityType As Type, ByVal token As Object) As LambdaExpression
			If token Is Nothing Then
				Throw New DataServiceException(401, "Unauthorized access.")
			End If
			If CStr(token) = "ok" Then
				Return CType(Function(o) True, Expression(Of Func(Of Object, Boolean)))
			Else
				Return CType(Function(o) False, Expression(Of Func(Of Object, Boolean)))
			End If
		End Function

		<WebGet> _
		Public Function GetAuthorizationCookie(ByVal name As String, ByVal password As String) As String
			Using uow As New UnitOfWork(Context.ObjectLayer)
				Dim contacts As List(Of Employees) = New XPQuery(Of Employees)(uow).Where(Function(i) i.LastName = name).ToList()
				Dim contact As Employees = contacts.SingleOrDefault()

				If contact Is Nothing Then
					Return String.Empty
				End If
				Dim ticket As New FormsAuthenticationTicket(1, contact.LastName, DateTime.Now, DateTime.Now.AddMinutes(30), True, contact.EmployeeID.ToString())

				Dim encryptedTicket As String = FormsAuthentication.Encrypt(ticket)
				Return encryptedTicket
			End Using
		End Function

		<WebGet> _
		Public Sub SetCategoriesFiltering(ByVal state As Boolean)
			context.CategoriesFiltering = state
		End Sub
	End Class
End Namespace
